dev/TEAL.md

```diff
@@ -0,0 +1,250 @@
+| `arg_3` | push LogicSig.Args[3] to stack |
+| `txn` | push field from current transaction to stack |
+| `gtxn` | push field to the stack from a transaction in the current transaction group |
+| `global` | push value from globals to stack |
+| `load` | copy a value from scratch space to the stack |
+| `store` | pop a value from the stack and store to scratch space |
+
+**Transaction Fields**
+
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | Sender | []byte |
+| 1 | Fee | uint64 |
+| 2 | FirstValid | uint64 |
+| 3 | FirstValidTime | uint64 |
+| 4 | LastValid | uint64 |
+| 5 | Note | []byte |
+| 6 | Lease | []byte |
+| 7 | Receiver | []byte |
+| 8 | Amount | uint64 |
+| 9 | CloseRemainderTo | []byte |
+| 10 | VotePK | []byte |
+| 11 | SelectionPK | []byte |
+| 12 | VoteFirst | uint64 |
+| 13 | VoteLast | uint64 |
+| 14 | VoteKeyDilution | uint64 |
+| 15 | Type | []byte |
+| 16 | TypeEnum | uint64 |
+| 17 | XferAsset | uint64 |
+| 18 | AssetAmount | uint64 |
+| 19 | AssetSender | []byte |
+| 20 | AssetReceiver | []byte |
+| 21 | AssetCloseTo | []byte |
+| 22 | GroupIndex | uint64 |
+| 23 | TxID | []byte |
+
+
+Additional details in the [opcodes document](TEAL_opcodes.md#txn) on the `txn` op.
+
+**Global Fields**
+
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | MinTxnFee | uint64 |
+| 1 | MinBalance | uint64 |
+| 2 | MaxTxnLife | uint64 |
+| 3 | ZeroAddress | []byte |
+| 4 | GroupSize | uint64 |
+
+
+
+### Flow Control
+
+| Op | Description |
+| --- | --- |
+| `err` | Error. Panic immediately. This is primarily a fencepost against accidental zero bytes getting compiled into programs. |
+| `bnz` | branch if value is not zero |
+| `pop` | discard value from stack |
+| `dup` | duplicate last value on stack |
+
+# Assembler Syntax
+
+The assembler parses line by line. Ops that just use the stack appear on a line by themselves. Ops that take arguments are the op and then whitespace and then any argument or arguments.
+
+"`//`" prefixes a line comment.
+
+## Constants and Pseudo-Ops
+
+A few pseudo-ops simplify writing code. `int` and `byte` and `addr` followed by a constant record the constant to a `intcblock` or `bytecblock` at the beginning of code and insert an `intc` or `bytec` reference where the instruction appears to load that value. `addr` parses an Algorand account address base32 and converts it to a regular bytes constant.
+
+`byte` constants are:
+```
+byte base64 AAAA...
+byte b64 AAAA...
+byte base64(AAAA...)
+byte b64(AAAA...)
+byte base32 AAAA...
+byte b32 AAAA...
+byte base32(AAAA...)
+byte b32(AAAA...)
+byte 0x0123456789abcdef...
+```
+
+`int` constants may be `0x` prefixed for hex, `0` prefixed for octal, or decimal numbers.
+
+`intcblock` may be explictly assembled. It will conflict with the assembler gathering `int` pseudo-ops into a `intcblock` program prefix, but may be used in code only has explicit `intc` references. `intcblock` should be followed by space separated int constants all on one line.
```

Comment:
good typo catch.
The assembler would write out an `intcblock` at the start of the program, and then the `intcblock` written into the program would run and replace the constants loaded by the first one. I don't want to support a mixed mode because I feel like it should be really rare for anyone to want to do that. All of our example scripts just use `int`/`byte`/`addr` and let the assembler manage `intcblock` and `bytecblock` and I expect that is what developers will do. The exception would be a compiled higher level language that does something like register allocation and thus arranges its own `intcblock`.